File: prelude/toolchains/android/src/com/facebook/buck/jvm/kotlin/buildtools/ExitCode.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is dual-licensed under either the MIT license found in the
 * LICENSE-MIT file in the root directory of this source tree or the Apache
 * License, Version 2.0 found in the LICENSE-APACHE file in the root directory
 * of this source tree. You may select, at your option, one of the
 * above-listed licenses.
 */

package com.facebook.buck.jvm.kotlin.buildtools

import com.facebook.buck.jvm.kotlin.kotlinc.Kotlinc
import org.jetbrains.kotlin.buildtools.api.CompilationResult

/**
 * Copy of [org.jetbrains.kotlin.cli.common.ExitCode]. It allows [BuildToolsKotlinc] to implement
 * the [Kotlinc] interface without depending on the Kotlin compiler directly.
 */
internal enum class ExitCode(val code: Int) {
  OK(0),
  COMPILATION_ERROR(1),
  INTERNAL_ERROR(2),
  OOM_ERROR(137)
}

internal val CompilationResult.toExitCode: ExitCode
  get() =
      when (this) {
        CompilationResult.COMPILATION_SUCCESS -> ExitCode.OK
        CompilationResult.COMPILATION_ERROR -> ExitCode.COMPILATION_ERROR
        CompilationResult.COMPILER_INTERNAL_ERROR -> ExitCode.INTERNAL_ERROR
        CompilationResult.COMPILATION_OOM_ERROR -> ExitCode.OOM_ERROR
        else -> error("Unexpected exit code: $this")
      }

File: app/buck2_error/src/exit_code.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is dual-licensed under either the MIT license found in the
 * LICENSE-MIT file in the root directory of this source tree or the Apache
 * License, Version 2.0 found in the LICENSE-APACHE file in the root directory
 * of this source tree. You may select, at your option, one of the
 * above-listed licenses.
 */

/// Common exit codes for buck with stronger semantic meanings
///
/// The exit code is u8 integer and has the following meanings for common exit codes.
/// - Success             : 0
/// - Uncategorized Error : 1
/// - Infra Error         : 2
/// - User Error          : 3
/// - Signal Interruption : 129-192 (128 + signal number)
#[derive(Clone, Copy, Debug)]
pub enum ExitCode {
    Success,
    UnknownFailure,
    InfraError,
    UserError,
    DaemonIsBusy,
    DaemonPreempted,
    Timeout,
    ConnectError,
    SignalInterrupt,
    ClientIoBrokenPipe,
    /// Test runner explicitly requested that this exit code be returned
    TestRunner(u8),
}

impl ExitCode {
    pub const fn exit_code(self) -> u32 {
        use ExitCode::*;
        match self {
            Success => 0,
            UnknownFailure => 1,
            InfraError => 2,
            UserError => 3,
            DaemonIsBusy => 4,
            DaemonPreempted => 5,
            Timeout => 6,
            ConnectError => 11,
            // Exit codes after 129 should be reserved for exits caused by signals,
            // which should be 128 + signal number.
            // However, SIGINT and SIGPIPE are reversed..
            ClientIoBrokenPipe => 130, // 128 + SIGINT(2)
            SignalInterrupt => 141,    // 128 + SIGPIPE(13)
            TestRunner(code) => code as u32,
        }
    }

    pub fn name(self) -> &'static str {
        use ExitCode::*;
        match self {
            Success => "SUCCESS",
            UnknownFailure => "UNKNOWN_FAILURE",
            InfraError => "INFRA_ERROR",
            UserError => "USER_ERROR",
            DaemonIsBusy => "DAEMON_IS_BUSY",
            DaemonPreempted => "DAEMON_PREEMPTED",
            Timeout => "TIMEOUT",
            ConnectError => "CONNECT_ERROR",
            ClientIoBrokenPipe => "CLIENT_IO_BROKEN_PIPE",
            SignalInterrupt => "SIGNAL_INTERRUPT",
            TestRunner(_) => "TEST_RUNNER",
        }
    }
}
